{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.ticker as tick\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import matplotlib.patches as pplt\n",
    "import matplotlib.cm as cm\n",
    "from matplotlib import lines\n",
    "import datetime\n",
    "import plotly.graph_objects as go\n",
    "from matplotlib import rcParams as rcp"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Setting variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#gene_name1 = input('Gene name 1: ')\n",
    "gene_name1 = 'brca1'\n",
    "#gene_name2 = input('Gene name 2 (if wanted, else press enter): ')\n",
    "gene_name2 = 'brca2'\n",
    "\n",
    "date = str(datetime.date.today())\n",
    "\n",
    "# joint file of gene\n",
    "df1 = pd.read_csv(gene_name1.lower() + '_joint_out.csv')\n",
    "dfs = [df1]\n",
    "gene_names = [gene_name1]\n",
    "if gene_name2 != '':\n",
    "    df2 = pd.read_csv(gene_name2.lower() + '_joint_out.csv')\n",
    "    dfs.append(df2)\n",
    "    gene_names.append(gene_name2)\n",
    "\n",
    "# SIFT_VEP and PolyPhen_VEP combined str and float values -> create one column for each\n",
    "for df in dfs:\n",
    "    df[['SIFT_VEP_str', 'SIFT_VEP_num']] = df['SIFT_VEP'].str.split('(', expand = True)\n",
    "    df['SIFT_VEP_num'] = df['SIFT_VEP_num'].str.strip(')').astype(float)\n",
    "    df[['PolyPhen_VEP_str', 'PolyPhen_VEP_num']] = df['PolyPhen_VEP'].str.split('(', expand = True)\n",
    "    df['PolyPhen_VEP_num'] = df['PolyPhen_VEP_num'].str.strip(')').astype(float)\n",
    "    df['Consequence_VEP'] = df['Consequence_VEP'].str.replace('_', ' ')\n",
    "    df['Consequence_VEP'] = df['Consequence_VEP'].str.replace(',', ', ')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# coding sequences as df\n",
    "coding_df1 = df1.loc[df1['Protein_position'].notnull()].copy()\n",
    "coding_dfs = [coding_df1]\n",
    "# unique positions (+ introns)\n",
    "all_pos1 = list(df1['POS'].unique())\n",
    "all_pos = [all_pos1]\n",
    "\n",
    "if gene_name2 != '':\n",
    "    coding_df2 = df2.loc[df2['Protein_position'].notnull()].copy()\n",
    "    coding_dfs.append(coding_df2)\n",
    "    all_pos2 = list(df2['POS'].unique())\n",
    "    all_pos.append(all_pos2)\n",
    "\n",
    "revs = []\n",
    "for df, cod in zip(dfs, coding_dfs):\n",
    "    if df.iloc[0]['Strand'] == '+':\n",
    "        cod.sort_values(by = ['POS'], ascending = True, ignore_index = True, inplace = True)\n",
    "        revs.append(False)\n",
    "    elif df.iloc[0]['Strand'] == '-':\n",
    "        cod.sort_values(by = ['POS'], ascending = False, ignore_index = True, inplace = True)\n",
    "        revs.append(True)\n",
    "    cod['Protein_position'] = cod['Protein_position'].astype(int)\n",
    "    cod['POS'] = cod['POS'].astype(int)\n",
    "\n",
    "# unique coding positions list\n",
    "codpos1 = list(coding_df1['POS'].unique())\n",
    "codpos = [codpos1]\n",
    "if gene_name2 != '':\n",
    "    codpos2 = list(coding_df2['POS'].unique())\n",
    "    codpos.append(codpos2)\n",
    "\n",
    "for pos,cpos,r in zip(all_pos, codpos, revs):\n",
    "    pos.sort(reverse = r)\n",
    "    cpos.sort(reverse = r)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# search for exon start/stop\n",
    "tuple_exons = []\n",
    "for df, cpos in zip(dfs, codpos):\n",
    "    exons_beg = []\n",
    "    exons_end = []\n",
    "    tuple_exons_in = []\n",
    "    if df.iloc[0]['Strand'] == '+':\n",
    "        for el in cpos:\n",
    "            if el-1 not in cpos:\n",
    "                exons_beg.append(el)\n",
    "            elif el+1 not in cpos:\n",
    "                exons_end.append(el)\n",
    "            else:\n",
    "                continue\n",
    "    elif df.iloc[0]['Strand'] == '-':\n",
    "        for el in cpos:\n",
    "            if el+1 not in cpos:\n",
    "                exons_beg.append(el)\n",
    "            elif el-1 not in cpos:\n",
    "                exons_end.append(el)\n",
    "            else:\n",
    "                continue\n",
    "    else:\n",
    "        pass\n",
    "    for x, y in zip(exons_beg, exons_end):\n",
    "        tuple_exons_in.append((x,y))\n",
    "    tuple_exons.append(tuple_exons_in)\n",
    "\n",
    "tuple_exons1 = tuple_exons[0]\n",
    "if len(tuple_exons) > 1:\n",
    "    tuple_exons2 = tuple_exons[1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Overview of gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# exons and introns for 1st gene\n",
    "fig,ax = plt.subplots(figsize=(40, 10))\n",
    "plt.ylim([-1, 1])\n",
    "plt.axis('off')\n",
    "fig.suptitle('Exons and introns of ' + gene_name1.upper() + ' gene in reading direction', fontsize = 30, fontweight = 'bold')\n",
    "plt.xlim(tuple_exons1[0][0], tuple_exons1[-1][1])\n",
    "if df1.iloc[0]['Strand'] == '+':\n",
    "    ax.add_patch(pplt.Rectangle((tuple_exons1[0][0], -.1), tuple_exons1[-1][1], .2, fc = 'blue', ec = 'blue', alpha = 1, label = 'Introns'))\n",
    "    ax.arrow(tuple_exons1[0][0], -0.7, abs(tuple_exons1[-1][1]-tuple_exons1[0][0])-1000, 0, head_width = 0.05, head_length = 1000, fc = 'k', ec = 'k')\n",
    "    for tup,l in zip(tuple_exons1, range(len(tuple_exons1))):\n",
    "        ax.add_patch(pplt.Rectangle((int(tup[0]), -.3), abs(int(tup[0])-int(tup[1])), .6, fc = 'red', ec = 'red', alpha = 1))\n",
    "        plt.text(int(tup[0]), -.4, str((int(l)+1)), size = 20)\n",
    "elif df1.iloc[0]['Strand'] == '-':\n",
    "    ax.add_patch(pplt.Rectangle((tuple_exons1[-1][1], -.1), tuple_exons1[0][0], .2, fc = 'blue', ec = 'blue', alpha = 1, label = 'Introns'))\n",
    "    ax.arrow(tuple_exons1[0][0], -0.7, -abs(tuple_exons1[-1][1]-tuple_exons1[0][0])+1000, 0, head_width = 0.05, head_length = 1000, fc = 'k', ec = 'k')\n",
    "    for tup,l in zip(tuple_exons1, range(len(tuple_exons1))):\n",
    "        ax.add_patch(pplt.Rectangle((int(tup[0]), -.3), abs(int(tup[0])-int(tup[1])), .6, fc = 'red', ec = 'red', alpha = 1))\n",
    "        plt.text(int(tup[0]+200), -.4, str((int(l)+1)), size = 20)\n",
    "\n",
    "plt.text(tuple_exons1[0][0], 0.7, 'Exons', c = 'red', size = 30)\n",
    "plt.text(tuple_exons1[0][0], 0.6, 'Introns', c = 'blue', size = 30)\n",
    "\n",
    "plt.text(tuple_exons1[0][0], -0.85, 'Reading Direction, \"' + str(df1.iloc[0]['Strand']) + '\" Strand', size = 30)\n",
    "plt.gcf().subplots_adjust(bottom = 0.05, top = 0.95, left = 0.05, right = 0.95);\n",
    "\n",
    "plt.savefig(date + '_' + gene_name1.lower() + '_exins_ov.pdf')\n",
    "plt.savefig(date + '_' + gene_name1.lower() + '_exins_ov.jpg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# exons and introns for 2nd gene\n",
    "if gene_name2 != '':\n",
    "    fig,ax = plt.subplots(figsize=(40, 10))\n",
    "    plt.ylim([-1, 1])\n",
    "    plt.axis('off')\n",
    "    fig.suptitle('Exons and introns of ' + gene_name2.upper() + ' gene in reading direction', fontsize = 30, fontweight = 'bold')\n",
    "    plt.xlim(tuple_exons2[0][0], tuple_exons2[-1][1])\n",
    "    if df2.iloc[0]['Strand'] == '+':\n",
    "        ax.add_patch(pplt.Rectangle((tuple_exons2[0][0], -.1), tuple_exons2[-1][1], .2, fc = 'blue', ec = 'blue', alpha = 1, label = 'Introns'))\n",
    "        ax.arrow(tuple_exons2[0][0], -0.7, abs(tuple_exons2[-1][1]-tuple_exons2[0][0])-1000, 0, head_width = 0.05, head_length = 1000, fc = 'k', ec = 'k')\n",
    "        for tup,l in zip(tuple_exons2, range(len(tuple_exons2))):\n",
    "            ax.add_patch(pplt.Rectangle((int(tup[0]), -.3), abs(int(tup[0])-int(tup[1])), .6, fc = 'red', ec = 'red', alpha = 1))\n",
    "            plt.text(int(tup[0]), -.4, str((int(l)+1)), size = 20)\n",
    "    elif df2.iloc[0]['Strand'] == '-':\n",
    "        ax.add_patch(pplt.Rectangle((tuple_exons2[-1][1], -.1), tuple_exons2[0][0], .2, fc = 'blue', ec = 'blue', alpha = 1, label = 'Introns'))\n",
    "        ax.arrow(tuple_exons2[0][0], -0.7, -abs(tuple_exons2[-1][1]-tuple_exons2[0][0])+1000, 0, head_width = 0.05, head_length = 1000, fc = 'k', ec = 'k')\n",
    "        for tup,l in zip(tuple_exons2, range(len(tuple_exons2))):\n",
    "            ax.add_patch(pplt.Rectangle((int(tup[0]), -.3), abs(int(tup[0])-int(tup[1])), .6, fc = 'red', ec = 'red', alpha = 1))\n",
    "            plt.text(int(tup[0]+200), -.4, str((int(l)+1)), size = 20)\n",
    "\n",
    "    plt.text(tuple_exons2[0][0], 0.7, 'Exons', c = 'red', size = 30)\n",
    "    plt.text(tuple_exons2[0][0], 0.6, 'Introns', c = 'blue', size = 30)\n",
    "\n",
    "    plt.text(tuple_exons2[0][0], -0.85, 'Reading Direction, \"' + str(df2.iloc[0]['Strand']) + '\" Strand', size = 30)\n",
    "    plt.gcf().subplots_adjust(bottom = 0.05, top = 0.95, left = 0.05, right = 0.95);\n",
    "\n",
    "    plt.savefig(date + '_' + gene_name2.lower() + '_exins_ov.pdf')\n",
    "    plt.savefig(date + '_' + gene_name2.lower() + '_exins_ov.jpg')\n",
    "else:\n",
    "    pass"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Overview of annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for pie chart with origin of variants, plus info of exon/intron\n",
    "# 1st gene\n",
    "sizes = []\n",
    "labels = []\n",
    "cols = []\n",
    "sizes_sg = []\n",
    "labels_sg = []\n",
    "cols_sg = []\n",
    "\n",
    "for df in dfs:\n",
    "    sizes.append([len(df[df['AC_gnomAD'].notnull() & df['ClinVar_gnomAD'].notnull()]), \n",
    "                  len(df[df['AC_gnomAD'].notnull() & df['ClinVar_gnomAD'].isnull()]), \n",
    "                  len(df[df['Cancer_Type_cBP'].notnull()]),\n",
    "                  len(df[df['Trans_Version_Flossies'].notnull()]), \n",
    "                  len(df[df['NCT_SAMPLE_NAME'].notnull()]), \n",
    "                 ])\n",
    "    labels.append(['GnomAD (+ ClinVar)',\n",
    "                    'GnomAD (w/o ClinVar)',\n",
    "                    'cBioPortal',\n",
    "                    'Flossies',\n",
    "                    'MASTER/NCT project',\n",
    "                  ])\n",
    "    cols.append(['red', 'orange', 'green', 'dodgerblue', 'magenta'])\n",
    "    sizes_sg.append([len(df[df['AC_gnomAD'].notnull() & df['ClinVar_gnomAD'].notnull() & df['EXON'].notnull()]), \n",
    "            len(df[df['AC_gnomAD'].notnull() & df['ClinVar_gnomAD'].notnull() & df['EXON'].isnull()]), \n",
    "            len(df[df['AC_gnomAD'].notnull() & df['ClinVar_gnomAD'].isnull() & df['EXON'].notnull()]), \n",
    "            len(df[df['AC_gnomAD'].notnull() & df['ClinVar_gnomAD'].isnull() & df['EXON'].isnull()]), \n",
    "            len(df[df['Cancer_Type_cBP'].notnull() & df['EXON'].notnull()]), \n",
    "            len(df[df['Cancer_Type_cBP'].notnull() & df['EXON'].isnull()]), \n",
    "            len(df[df['Trans_Version_Flossies'].notnull() & df['EXON'].notnull()]), \n",
    "            len(df[df['Trans_Version_Flossies'].notnull() & df['EXON'].isnull()]), \n",
    "            len(df[df['NCT_SAMPLE_NAME'].notnull() & df['EXON'].notnull()]),\n",
    "            len(df[df['NCT_SAMPLE_NAME'].notnull() & df['EXON'].isnull()]),\n",
    "           ])\n",
    "    labels_sg.append(['E', 'I', 'E', 'I', 'E', 'I', 'E', 'I', 'E', 'I'])\n",
    "    cols_sg.append(['lightcoral', 'lightsalmon', 'gold', 'bisque', 'chartreuse', 'palegreen', \n",
    "          'deepskyblue', 'lightskyblue', 'violet', 'lightpink'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.subplots(figsize=(16, 8))\n",
    "# pie plot in coding sequence + 75 bp padding for gene 1, exon vs. intron\n",
    "if gene_name2 != '':\n",
    "    ax1 = plt.subplot(1,2,1)\n",
    "else:\n",
    "    ax1 = plt.subplot(1,1,1)\n",
    "ax1.set_title('Origin of variants in ' + gene_name1.upper() + ' gene', fontsize = 15, fontweight = 'bold')\n",
    "ax1.pie(sizes[0], labels = labels[0], autopct = '%1.1f%%', startangle = 180, pctdistance = 0.85, \n",
    "        colors = cols[0], counterclock = False, radius = 1, wedgeprops = dict(width = 0.3))\n",
    "ax1.pie(sizes_sg[0], autopct = '%1.1f%%', startangle = 180, pctdistance = 0.8, \n",
    "            labels = labels_sg[0], labeldistance = 0.45, colors = cols_sg[0], \n",
    "            counterclock = False, radius = 0.7, wedgeprops = dict(width = 0.3));\n",
    "\n",
    "# pie plot in coding sequence + 75 bp padding for gene 2, exon vs. intron\n",
    "if gene_name2 != '':\n",
    "    ax2 = plt.subplot(1,2,2)\n",
    "    ax2.set_title('Origin of variants in ' + gene_name2.upper() + ' gene', fontsize = 15, fontweight = 'bold')\n",
    "    ax2.pie(sizes[1], labels = labels[1], autopct = '%1.1f%%', startangle = 180, pctdistance = 0.85, \n",
    "                colors = cols[1], counterclock = False, radius = 1, wedgeprops = dict(width = 0.3))\n",
    "    ax2.pie(sizes_sg[1], autopct = '%1.1f%%', startangle = 180, pctdistance = 0.8, \n",
    "                labels = labels_sg[1], labeldistance = 0.45, colors = cols_sg[1], \n",
    "                counterclock = False, radius = 0.7, wedgeprops = dict(width = 0.3));\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(date + '_' + gene_name1.lower() + '_' + gene_name2.lower() + '_orig_vars.pdf')\n",
    "    plt.savefig(date + '_' + gene_name1.lower() + '_' + gene_name2.lower() + '_orig_vars.jpg')\n",
    "else:\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(date + '_' + gene_name1.lower() + '_orig_vars.pdf')\n",
    "    plt.savefig(date + '_' + gene_name1.lower() + '_orig_vars.jpg')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Overview of ClinVar/VEP impact"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for pathogenicity by ClinVar, all, exons, introns\n",
    "path1_val = df1['ClinVar_gnomAD'][df1['ClinVar_gnomAD'].notnull()].value_counts()\n",
    "path1_val_df = pd.DataFrame(path1_val)\n",
    "path1_val_ex1_df = pd.DataFrame(df1['ClinVar_gnomAD'][(df1['ClinVar_gnomAD'].notnull()) &\n",
    "                                      (df1['EXON'].notnull())].value_counts())\n",
    "path1_val_in1_df = pd.DataFrame(df1['ClinVar_gnomAD'][(df1['ClinVar_gnomAD'].notnull()) &\n",
    "                                      (df1['INTRON'].notnull())].value_counts())\n",
    "# for impact by VEP\n",
    "imp1 = df1['Consequence_VEP'][df1['Consequence_VEP'].notnull()].value_counts()\n",
    "\n",
    "# for gene 2\n",
    "if gene_name2 != '':\n",
    "    path2_val = df2['ClinVar_gnomAD'][df2['ClinVar_gnomAD'].notnull()].value_counts()\n",
    "    path2_val_df = pd.DataFrame(path2_val)\n",
    "    path2_val_ex2_df = pd.DataFrame(df2['ClinVar_gnomAD'][(df2['ClinVar_gnomAD'].notnull()) &\n",
    "                                          (df2['EXON'].notnull())].value_counts())\n",
    "    path2_val_in2_df = pd.DataFrame(df2['ClinVar_gnomAD'][(df2['ClinVar_gnomAD'].notnull()) &\n",
    "                                          (df2['INTRON'].notnull())].value_counts())\n",
    "    imp2 = df2['Consequence_VEP'][df2['Consequence_VEP'].notnull()].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for el in [path1_val_df, path1_val_ex1_df, path1_val_in1_df, path2_val_df, path2_val_ex2_df, path2_val_in2_df]:\n",
    "    el['index_1'] = el.index\n",
    "all_pat1_df = path1_val_df.merge(path1_val_ex1_df, on = 'index_1', suffixes = ['_all1', '_ex1']\n",
    "                  ).merge(path1_val_in1_df, on = 'index_1')\n",
    "all_pat1_df = all_pat1_df.rename(columns = {'ClinVar_gnomAD':'ClinVar_gnomAD_in1'})\n",
    "all_pat2_df = path2_val_df.merge(path2_val_ex2_df, on = 'index_1', suffixes = ['_all2', '_ex2']\n",
    "                  ).merge(path2_val_in2_df, on = 'index_1')\n",
    "all_pat2_df = all_pat2_df.rename(columns = {'ClinVar_gnomAD':'ClinVar_gnomAD_in2'})\n",
    "# --> with these stacked bar plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for impact in introns/exons\n",
    "imp_in1 = df1['Consequence_VEP'][(df1['Consequence_VEP'].notnull()) & \n",
    "                                 (df1['INTRON'].notnull())].value_counts()\n",
    "imp_ex1 = df1['Consequence_VEP'][(df1['Consequence_VEP'].notnull()) &\n",
    "                                 (df1['EXON'].notnull())].value_counts()\n",
    "\n",
    "col1_imp1 = []\n",
    "for el in imp1.index:\n",
    "    if el in imp_in1.index and not el in imp_ex1.index:\n",
    "        col1_imp1.append('orange')\n",
    "    elif el in imp_ex1.index and not el in imp_in1.index:\n",
    "        col1_imp1.append('dodgerblue')\n",
    "    elif el in imp_in1.index and el in imp_ex1.index:\n",
    "        col1_imp1.append('magenta')\n",
    "\n",
    "if gene_name2 != '':\n",
    "    imp_in2 = df2['Consequence_VEP'][(df2['Consequence_VEP'].notnull()) &\n",
    "                                     (df2['INTRON'].notnull())].value_counts()\n",
    "    imp_ex2 = df2['Consequence_VEP'][(df2['Consequence_VEP'].notnull()) &\n",
    "                                     (df2['EXON'].notnull())].value_counts()\n",
    "    col2_imp2 = []\n",
    "    for el in imp2.index:\n",
    "        if el in imp_in2.index and not el in imp_ex2.index:\n",
    "            col2_imp2.append('orange')\n",
    "        elif el in imp_ex2.index and not el in imp_in2.index:\n",
    "            col2_imp2.append('dodgerblue')\n",
    "        elif el in imp_in2.index and el in imp_ex2.index:\n",
    "            col2_imp2.append('magenta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.subplots(figsize=(16, 16))\n",
    "if gene_name2 != '':\n",
    "    ax1 = plt.subplot(2,2,1)\n",
    "    ax3 = plt.subplot(2,2,3)\n",
    "else:\n",
    "    ax1 = plt.subplot(2,1,1)\n",
    "    ax3 = plt.subplot(2,1,2)\n",
    "\n",
    "blue_leg = pplt.Patch(color = 'dodgerblue', label = 'in Exons')\n",
    "or_leg = pplt.Patch(color = 'orange', label = 'in Introns')\n",
    "mag_leg = pplt.Patch(color = 'magenta', label = 'in both')\n",
    "\n",
    "# impact of variants gene 1\n",
    "ax1.set_title('Impact of variants in ' + gene_name1.upper() + ' gene (by VEP)', fontsize = 15, fontweight = 'bold')\n",
    "ax1.bar(imp1.index, imp1.values, color = col1_imp1)\n",
    "ax1.set_xticklabels(imp1.index, rotation = 45, ha = 'right')\n",
    "ax1.set_ylabel('Variant count (log scale)')\n",
    "ax1.set_yscale('log')\n",
    "ax1.legend(handles = [blue_leg, or_leg, mag_leg])\n",
    "\n",
    "# pathogenicity gene 1\n",
    "ax3.set_title('GnomAD ClinVar annotations in ' + gene_name1.upper() + ' gene', fontsize = 15, fontweight = 'bold')\n",
    "ax3.bar(all_pat1_df['index_1'], all_pat1_df['ClinVar_gnomAD_ex1'], \n",
    "       label = 'exons', color = 'dodgerblue')\n",
    "ax3.bar(all_pat1_df['index_1'], all_pat1_df['ClinVar_gnomAD_in1'],\n",
    "        label = 'introns', bottom = all_pat1_df['ClinVar_gnomAD_ex1'], \n",
    "        color = 'orange')\n",
    "ax3.set_xticklabels(path1_val.index, rotation = 45, ha = 'right')\n",
    "ax3.set_ylabel('Variant count')\n",
    "#ax3.set_yscale('log')\n",
    "ax3.legend(handles = [blue_leg, or_leg])\n",
    "\n",
    "if gene_name2 != '':\n",
    "    # impact of variants gene 2\n",
    "    ax2 = plt.subplot(2,2,2, sharey = ax1)\n",
    "    ax2.set_title('Impact of variants in ' + gene_name2.upper() + ' gene (by VEP)', fontsize = 15, fontweight = 'bold')\n",
    "    ax2.bar(imp2.index, imp2.values, color = col2_imp2)\n",
    "    ax2.set_xticklabels(imp2.index, rotation = 45, ha = 'right')\n",
    "    ax2.set_ylabel('Variant count (log scale)')\n",
    "    ax2.set_yscale('log')\n",
    "    ax2.legend(handles = [blue_leg, or_leg, mag_leg])\n",
    "    # pathogenicity gene 2\n",
    "    ax4 = plt.subplot(2,2,4, sharey = ax3)\n",
    "    ax4.set_title('GnomAD ClinVar annotations in ' + gene_name2.upper() + ' gene', fontsize = 15, fontweight = 'bold')\n",
    "    ax4.bar(all_pat2_df['index_1'], all_pat2_df['ClinVar_gnomAD_ex2'], \n",
    "       label = 'exons', color = 'dodgerblue')\n",
    "    ax4.bar(all_pat2_df['index_1'], all_pat2_df['ClinVar_gnomAD_in2'],\n",
    "        label = 'introns', bottom = all_pat2_df['ClinVar_gnomAD_ex2'], \n",
    "        color = 'orange')\n",
    "    ax4.set_xticklabels(path2_val.index, rotation = 45, ha = 'right')\n",
    "    ax4.set_ylabel('Variant count')\n",
    "    #ax4.set_yscale('log')\n",
    "    ax4.legend(handles = [blue_leg, or_leg]);\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(date + '_' + gene_name1.lower() + '_' + gene_name2.lower() + '_vars_cons_impact.pdf')\n",
    "    plt.savefig(date + '_' + gene_name1.lower() + '_' + gene_name2.lower() + '_vars_cons_impact.jpg')\n",
    "else:\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(date + '_' + gene_name1.lower() + '_vars_cons_impact.pdf')\n",
    "    plt.savefig(date + '_' + gene_name1.lower() + '_vars_cons_impact.jpg')\n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Consequences of pathogenic variants, ClinVar of most common variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# consequences of pathogenic variants\n",
    "cons_pat1 = df1['Consequence_VEP'][df1['ClinVar_gnomAD'] == 'Pathogenic'].value_counts()\n",
    "# clinvar annotations of most common consequence\n",
    "clinv_miss1 = df1['ClinVar_gnomAD'][df1['Consequence_VEP'] == 'missense variant'].value_counts()\n",
    "\n",
    "if gene_name2 != '':\n",
    "    cons_pat2 = df2['Consequence_VEP'][df2['ClinVar_gnomAD'] == 'Pathogenic'].value_counts()\n",
    "    clinv_miss2 = df2['ClinVar_gnomAD'][df2['Consequence_VEP'] == 'missense variant'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.subplots(figsize=(16, 16))\n",
    "if gene_name2 != '':\n",
    "    ax5 = plt.subplot(2,2,1)\n",
    "    ax3 = plt.subplot(2,2,3)\n",
    "else:\n",
    "    ax5 = plt.subplot(2,1,1)\n",
    "    ax3 = plt.subplot(2,1,2)\n",
    "# consequences of pathogenic variants gene 1\n",
    "ax5.set_title('Consequences of pathogenic variants in ' + gene_name1.upper() + ' gene (by VEP)', fontsize = 15, fontweight = 'bold')\n",
    "ax5.bar(cons_pat1.index, cons_pat1.values)\n",
    "ax5.set_xticklabels(cons_pat1.index, rotation = 45, ha = 'right')\n",
    "ax5.set_ylabel('Variant count (log scale)')\n",
    "ax5.set_yscale('log')\n",
    "# clinvar annotations in most common consequence in gene 1\n",
    "ax3.set_title('ClinVar annotations of missense variants in ' + gene_name1.upper() + ' gene', fontsize = 15, fontweight = 'bold')\n",
    "ax3.bar(clinv_miss1.index, clinv_miss1.values)\n",
    "ax3.set_xticklabels(clinv_miss1.index, rotation = 45, ha = 'right')\n",
    "ax3.set_ylabel('Variant count (log scale)')\n",
    "ax3.set_yscale('log')\n",
    "\n",
    "if gene_name2 != '':\n",
    "    # consequences of pathogenic variants gene 2\n",
    "    ax6 = plt.subplot(2,2,2, sharey = ax5)\n",
    "    ax6.set_title('Consequences of pathogenic variants in ' + gene_name2.upper() + ' gene (by VEP)', fontsize = 15, fontweight = 'bold')\n",
    "    ax6.bar(cons_pat2.index, cons_pat2.values)\n",
    "    ax6.set_xticklabels(cons_pat2.index, rotation = 45, ha = 'right')\n",
    "    ax6.set_ylabel('Variant count (log scale)')\n",
    "    ax6.set_yscale('log')\n",
    "    # clinvar annotations in most common consequence in gene 2\n",
    "    ax4 = plt.subplot(2,2,4, sharey = ax3)\n",
    "    ax4.set_title('ClinVar annotations of missense variants in ' + gene_name2.upper() + ' gene', fontsize = 15, fontweight = 'bold')\n",
    "    ax4.bar(clinv_miss2.index, clinv_miss2.values)\n",
    "    ax4.set_xticklabels(clinv_miss2.index, rotation = 45, ha = 'right')\n",
    "    ax4.set_ylabel('Variant count (log scale)')\n",
    "    ax4.set_yscale('log')\n",
    "    plt.tight_layout();\n",
    "    plt.savefig(date + '_' + gene_name1.lower() + '_' + gene_name2.lower() + '_path_miss.pdf')\n",
    "    plt.savefig(date + '_' + gene_name1.lower() + '_' + gene_name2.lower() + '_path_miss.jpg')\n",
    "else:\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(date + '_' + gene_name1.lower() + '_path_miss.pdf')\n",
    "    plt.savefig(date + '_' + gene_name1.lower() + '_path_miss.jpg')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Location of as pathogenic characterized variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "fig = plt.subplots(figsize=(16, 8))\n",
    "if gene_name2 != '':\n",
    "    ax1 = plt.subplot(2,1,1)\n",
    "else:\n",
    "    ax1 = plt.subplot(1,1,1)\n",
    "# location of pathogenic variants with frameshift/stop gained gene 1\n",
    "ax1.set_title('Location of pathogenic variants in ' + gene_name1.upper() + ' gene', fontsize = 15, fontweight = 'bold')\n",
    "ax1.scatter(x = df1['POS'][(df1['ClinVar_gnomAD'] == 'Pathogenic') &\n",
    "                           (df1['Consequence_VEP'] == 'frameshift variant')], \n",
    "            y = df1['CADD_PHRED_VEP'][(df1['ClinVar_gnomAD'] == 'Pathogenic') &\n",
    "                                       (df1['Consequence_VEP'] == 'frameshift variant')], \n",
    "            alpha = 0.5, label = 'frameshift variant')\n",
    "ax1.scatter(x = df1['POS'][(df1['ClinVar_gnomAD'] == 'Pathogenic') &\n",
    "                           (df1['Consequence_VEP'] == 'stop gained')], \n",
    "            y = df1['CADD_PHRED_VEP'][(df1['ClinVar_gnomAD'] == 'Pathogenic') &\n",
    "                                      (df1['Consequence_VEP'] == 'stop gained')], \n",
    "            alpha = 0.5, label = 'stop gained')\n",
    "ax1.set_xlabel('Nucleic position at ' + gene_name1.upper(), fontsize = 13)\n",
    "ax1.set_ylabel('CADD phred score (VEP)', fontsize = 13)\n",
    "ax1.ticklabel_format(style = 'plain')\n",
    "ax1.get_xaxis().set_major_formatter(tick.StrMethodFormatter('{x:,.0f}'))\n",
    "plt.legend(bbox_to_anchor = (1, 1.12), ncol = 2)\n",
    "\n",
    "\n",
    "if gene_name2 != '':\n",
    "    # consequences of pathogenic variants gene 2\n",
    "    ax2 = plt.subplot(2,1,2, sharey = ax1)\n",
    "    ax2.set_title('Location of pathogenic variants in ' + gene_name2.upper() + ' gene', fontsize = 15, fontweight = 'bold')\n",
    "    ax2.scatter(x = df2['POS'][(df2['ClinVar_gnomAD'] == 'Pathogenic') &\n",
    "                               (df2['Consequence_VEP'] == 'frameshift variant')], \n",
    "                y = df2['CADD_PHRED_VEP'][(df2['ClinVar_gnomAD'] == 'Pathogenic') &\n",
    "                                           (df2['Consequence_VEP'] == 'frameshift variant')], \n",
    "                alpha = 0.5, label = 'frameshift variant')\n",
    "    ax2.scatter(x = df2['POS'][(df2['ClinVar_gnomAD'] == 'Pathogenic') &\n",
    "                               (df2['Consequence_VEP'] == 'stop gained')], \n",
    "                y = df2['CADD_PHRED_VEP'][(df2['ClinVar_gnomAD'] == 'Pathogenic') &\n",
    "                                          (df2['Consequence_VEP'] == 'stop gained')], \n",
    "                alpha = 0.5, label = 'stop gained')\n",
    "    ax2.set_xlabel('Nucleic position at ' + gene_name2.upper(), fontsize = 13)\n",
    "    ax2.set_ylabel('CADD phred score (VEP)', fontsize = 13)\n",
    "    ax2.ticklabel_format(style = 'plain')\n",
    "    ax2.get_xaxis().set_major_formatter(tick.StrMethodFormatter('{x:,.0f}'))\n",
    "    plt.legend(bbox_to_anchor = (1, 1.12), ncol = 2)\n",
    "    plt.tight_layout();\n",
    "    plt.savefig(date + '_' + gene_name1.lower() + '_' + gene_name2.lower() + '_loc_pat.pdf')\n",
    "    plt.savefig(date + '_' + gene_name1.lower() + '_' + gene_name2.lower() + '_loc_pat.jpg')\n",
    "else:\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(date + '_' + gene_name1.lower() + '_loc_pat.pdf')\n",
    "    plt.savefig(date + '_' + gene_name1.lower() + '_loc_pat.jpg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "divide variants into groups of deleteriousness\n",
    "\n",
    "set thresholds for deleteriosness of variants\n",
    "- whole df: 42623 variants\n",
    "- AF_gnomAD < 0.001 (allele frequency, 3600 variants of 3670)\n",
    "- CADD_PHRED_VEP > 15 (rank of variant to substitution in whole human reference genome, 15571 variants of 42348)\n",
    "- Ada_score_VEP > 0.6 (splicing effect score, 710 variants of 1884)\n",
    "- SIFT_VEP_num < 0.05 (resulting changed protein function by aa change, 10553 of 22772)\n",
    "- PolyPhen_VEP_num > 0.6 (resulting changed protein structure/function by aa change, 6073 of 22772)\n",
    "- Impact_VEP == 'HIGH'\n",
    "- ClinVar_gnomAD == 'Pathogenic', 'Pathogenic/Likely pathogenic', 'Conflicting interpretations of pathogenicity', 'not provided', 'Uncertain significance'\n",
    "\n",
    "results:\n",
    "- for everything: no result\n",
    "- if no Ada_score (for splice effects) --> one result: start_lost (of 9x 'start lost')\n",
    "- if add. no Impact_VEP --> 366 results for 'cplt_path_group'\n",
    "\n",
    "\n",
    "'''\n",
    "\n",
    "# variants completely categorized as deleterious\n",
    "cplt_path_group = []\n",
    "cplt_ben_group = []\n",
    "for df in dfs:\n",
    "    cplt_path_group.append(df[(df['AF_gnomAD'] < 0.001) &\n",
    "                              (df['SIFT_VEP_num'] < 0.05) &\n",
    "                              (df['CADD_PHRED_VEP'] > 15) &\n",
    "                              (df['PolyPhen_VEP_num'] > 0.6) &\n",
    "                              (df['ClinVar_gnomAD'].isin(['Pathogenic', 'Pathogenic/Likely pathogenic', \n",
    "                                                          'Conflicting interpretations of pathogenicity', \n",
    "                                                          'not provided', 'Uncertain significance']))])\n",
    "    cplt_ben_group.append(df[(df['AF_gnomAD'] >= 0.001) &\n",
    "                             (df['SIFT_VEP_num'] >= 0.05) &\n",
    "                             (df['CADD_PHRED_VEP'] <= 15) &\n",
    "                             (df['PolyPhen_VEP_num'] <= 0.6)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# nucleic positions of variants in cplt pathogen group + ClinVar annotation\n",
    "\n",
    "fig = plt.subplots(figsize=(16, 16))\n",
    "\n",
    "if gene_name2 != '':\n",
    "    ax1 = plt.subplot(2,1,1)\n",
    "    ax2 = plt.subplot(2,1,2)\n",
    "    axes = [ax1, ax2]\n",
    "else:\n",
    "    ax1 = plt.subplot(1,1,1)\n",
    "    axes = [ax1]\n",
    "    \n",
    "for p, g, a in zip(cplt_path_group, gene_names, axes):\n",
    "    a.scatter(p['POS'], p['ClinVar_gnomAD'])\n",
    "    a.set_title('Nucleic position and ClinVar categorization of most deleterious variants in ' + \n",
    "                g.upper() + ' gene')\n",
    "    a.set_xlabel('Nucleic position at ' + g.upper() + ' gene')\n",
    "#plt.xlabel('Nucleic position at ' + gene_name[0].upper() + ' and ' + gene_name[1].upper() + ' genes')\n",
    "#plt.title('Nucleic position and ClinVar categorization of most deleterious variants in ' + gene_names[0].upper() + ' and ' + gene_names[0].upper() + ' gene');"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Deleteriousness of variants"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GnomAD Allele Frequency vs. CADD phred score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# allele frequency and cadd score of variants\n",
    "fig = plt.figure(figsize = (10,10))\n",
    "for df, g in zip(dfs, gene_names):\n",
    "    plt.scatter(x = df['CADD_PHRED_VEP'], y = df['AF_gnomAD'], label = g.upper(), alpha = 0.5)\n",
    "plt.yscale('log')\n",
    "plt.ylabel('Allele Frequency by gnomAD (log scale)')\n",
    "plt.xlabel('CADD phred score (VEP)')\n",
    "plt.legend()\n",
    "\n",
    "tit_g = (' + ').join(gene_names).upper()\n",
    "plt.title(label = ('Allele Frequency (gnomAD) to CADD phred score (VEP) in ' + str(tit_g)), fontsize = 15, fontweight = 'bold');\n",
    "if len(gene_names) > 1:\n",
    "    plt.savefig(date + '_' + gene_names[0] + '_' + gene_names[1] + '_cadd_af.pdf')\n",
    "    plt.savefig(date + '_' + gene_names[0] + '_' + gene_names[1] + '_cadd_af.jpg')\n",
    "else:\n",
    "    plt.savefig(date + '_' + gene_names[0] + '_cadd_af.pdf')\n",
    "    plt.savefig(date + '_' + gene_names[0] + '_cadd_af.jpg')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Division in groups for prediction of deleteriousness"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# SIFT_VEP vs. PolyPhen_VEP\n",
    "fig = plt.subplots(figsize = (10,10))\n",
    "\n",
    "#for df, g in zip(dfs, gene_names):\n",
    "ax1 = plt.subplot(2,1,1)\n",
    "ax1.set_title('PolyPhen score (VEP) to SIFT score (VEP) in ' + gene_name1.upper() + ' gene', \n",
    "              fontsize = 15, fontweight = 'bold')\n",
    "ax1.scatter(x = df1['SIFT_VEP_num'], y = df1['PolyPhen_VEP_num'], \n",
    "                  label = gene_name1.upper(), alpha = 0.5)\n",
    "ax1.set_ylabel('SIFT score by VEP')\n",
    "ax1.set_xlabel('PolyPhen score by VEP')\n",
    "#plt.legend()\n",
    "\n",
    "ax2 = plt.subplot(2,1,2, sharey = ax1)\n",
    "ax2.set_title('PolyPhen score (VEP) to SIFT score (VEP) in ' + gene_name2.upper() + ' gene', \n",
    "              fontsize = 15, fontweight = 'bold')\n",
    "ax2.scatter(x = df2['SIFT_VEP_num'], y = df2['PolyPhen_VEP_num'], \n",
    "                  label = gene_name2.upper(), alpha = 0.5, c = 'orange')\n",
    "ax2.set_ylabel('SIFT score by VEP')\n",
    "ax2.set_xlabel('PolyPhen score by VEP')\n",
    "plt.tight_layout()\n",
    "\n",
    "plt.savefig(date + '_' + gene_names[0] + '_' + gene_names[1] + '_cadd_af.pdf')\n",
    "plt.savefig(date + '_' + gene_names[0] + '_' + gene_names[1] + '_cadd_af.jpg')\n",
    "\n",
    "#if len(gene_names) > 1:\n",
    "    #plt.savefig(r'figures/' + date + '_' + gene_names[0] + '_' + gene_names[1] + '_cadd_af.pdf')\n",
    "    #plt.savefig(r'figures/' + date + '_' + gene_names[0] + '_' + gene_names[1] + '_cadd_af.jpg')\n",
    "#else:\n",
    "    #plt.savefig(r'figures/' + date + '_' + gene_names[0] + '_cadd_af.pdf')\n",
    "    #plt.savefig(r'figures/' + date + '_' + gene_names[0] + '_cadd_af.jpg')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Stop gain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Where are variants with gained stop codons?\n",
    "Which variants are how annotated in ClinVar, which in cBioPortal?\n",
    "'''\n",
    "\n",
    "# for stop gain --> in which exons are variants with gained stop codons, which are how annotated in ClinVar, which in cBioPortal\n",
    "plt.figure(figsize = (16,10))\n",
    "sns.scatterplot(x = df1['POS'][df1['Consequence_VEP'] == 'stop gained'], y = df1['CADD_PHRED_VEP'][df1['Consequence_VEP'] == 'stop gained'], hue = df1['ClinVar_gnomAD'][df1['Consequence_VEP'] == 'stop gained'].fillna('Not in GnomAD'), style = df1['MS_cBP'][df1['Consequence_VEP'] == 'stop gained'].fillna('Not in cBioPortal'))\n",
    "plt.title('Stop Gained Variants in ' + gene_name1.upper() + ' with ClinVar annotations of GnomAD')\n",
    "plt.xlabel('Nucleic position of ' + gene_name1.upper())\n",
    "plt.ylabel('CADD phred score (by VEP)')\n",
    "plt.legend(bbox_to_anchor = (1.01, 1), borderaxespad = 0);\n",
    "plt.savefig(date + '_' + gene_name1 + '_stop_gained.pdf')\n",
    "plt.savefig(date + '_' + gene_name1 + '_stop_gained.jpg')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
